text/0062-memory-data.md

```diff
@@ -0,0 +1,116 @@
+
+The code to create a memory changes from:
+
+```py
+m.submodules.memory = memory = Memory(shape=..., depth=..., init=...)
+port = memory.read_port(...)
+```
+
+to:
+
+```py
+# in __init__
+self.mem_data = MemoryData(shape=..., depth=..., init=...)
+# in elaborate
+m.submodules.memory = memory = Memory(self.mem_data)
+port = memory.read_port(...)
+```
+
+The `my_component.mem_data` object can then be used in simulation to read and write memory:
+
+```py
+addr = 0x1234
+row = sim.get(mem_data[addr])
+row += 1
+sim.set(mem_data[addr], row)
+```
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Two new classes are added:
+
+- `amaranth.hdl.MemoryData(*, shape: ShapeLike, depth: int, init: Iterable[int | Any], name=None)`: represents a memory's data storage. `name`, if not specified, defaults to the variable name used to store the `MemoryData`, like it does for `Signal`.
+  - `__getitem__(addr: int) -> MemoryData.Row`: creates a `MemoryData.Row` object
+- `amaranth.hdl.MemoryData.Row`: represents a single row of `MemoryData`, has no public constructor, can be passed to simulator functions
+
+The `MemoryData` class allows access to its constructor arguments via read-only properties.
+
+The `lib.memory.Memory.Init` class is moved to `amaranth.hdl.MemoryData.Init`. It is used for the `init` property of `MemoryData`.
+
+The `Memory` constructor is changed to:
+
+- `amaranth.lib.memory.Memory(data: MemoryData = None, *, shape=None, depth=None, init=None, name=None)`
+
+  - either `data`, or all three of `shape`, `depth`, `init` need to be provided, but not both
+  - if `data` is provided, it is used directly, and stored
+  - if `shape`, `depth`, `init` (and possibly `name`) are provided, they are used to create a `MemoryData`, which is then stored
+
+The `MemoryData` object is accessible via a new read-only `data` property on `Memory`. The existing `shape`, `depth`, `init` properties become aliases for `data.shape`, `data.depth`, `data.init`.
+
+The `sim.memory_read` and `sim.memory_write` methods proposed by RFC 36 are removed. Instead, the following simulation context methods are amended to accept `MemoryData.Row` objects in addition to `Value` and `ValueCastable`:
+
+- `get()`
+- `set()` first argument
+- `changed()`
+- (proposed by [amaranth-lang/rfcs#64](https://github.com/amaranth-lang/rfcs/pull/64)) `tick().sample()`
+
+If `MemoryData.shape` is a `ShapeCastable`, the values are converted using `ShapeCastable.const` / `ShapeCastable.from_bits`. Otherwise, they need to be integers.
+
+`MemoryData` and `MemoryData.Row` instances can be added to the `traces` argument when writing a VCD file with the simulator.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+The ability to do a masked write introduced by RFC 36 is removed. It could be reintroduced (by adding `mask` to `set`, or by somehow accepting `sim.set(mem_data[row][sli:ce], ...)`), but it adds complexity and is redundant with `get` / modify / `set`.
```

Comment:
>  is redundant with `get` / modify / `set`.

Unfortunately this is not true. Several `add_process` processes writing the same signal will race if they all do an RMW, but not race if they do a masked write. (The same would *seem* to not be relevant for signals because you can write only a slice of a signal, but is actually broken in the same way because of our implementation.)